src/coprocessor/dag/rpn_expr/types.rs

```diff
@@ -0,0 +1,695 @@
+                }
+                rpn_nodes.push(RpnExpressionNode::Fn { func, field_type });
+            }
+            ExprType::ColumnRef => {
+                let offset = number::decode_i64(&mut def.get_val()).map_err(|_| {
+                    Error::Other(box_err!(
+                        "Unable to decode column reference offset from the request"
+                    ))
+                })? as usize;
+                rpn_nodes.push(RpnExpressionNode::TableColumnRef { offset, field_type });
+            }
+            t => return Err(box_err!("Unexpected ExprType {:?}", t)),
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tipb::expression::ScalarFuncSig;
+
+    /// An RPN function for test. It accepts 1 int argument, returns the value in float.
+    #[derive(Debug, Clone, Copy)]
+    struct FnA;
+
+    impl_template_fn! { 1 arg @ FnA }
+
+    /// An RPN function for test. It accepts 2 float arguments, returns their sum in int.
+    #[derive(Debug, Clone, Copy)]
+    struct FnB;
+
+    impl_template_fn! { 2 arg @ FnB }
+
+    /// An RPN function for test. It accepts 3 int arguments, returns their sum in int.
+    #[derive(Debug, Clone, Copy)]
+    struct FnC;
+
+    impl_template_fn! { 3 arg @ FnC }
+
+    /// An RPN function for test. It accepts 3 float arguments, returns their sum in float.
+    #[derive(Debug, Clone, Copy)]
+    struct FnD;
+
+    impl_template_fn! { 3 arg @ FnD }
+
+    /// For testing `append_rpn_nodes_recursively`. It accepts protobuf function sig enum, which
+    /// cannot be modified by us in tests to support FnA ~ FnD. So let's just hard code some
+    /// substitute.
+    fn fn_mapper(value: ScalarFuncSig) -> Result<Box<dyn RpnFunction>> {
+        // FnA: CastIntAsInt
+        // FnB: CastIntAsReal
+        // FnC: CastIntAsString
+        // FnD: CastIntAsDecimal
+        match value {
+            ScalarFuncSig::CastIntAsInt => Ok(Box::new(FnA)),
+            ScalarFuncSig::CastIntAsReal => Ok(Box::new(FnB)),
+            ScalarFuncSig::CastIntAsString => Ok(Box::new(FnC)),
+            ScalarFuncSig::CastIntAsDecimal => Ok(Box::new(FnD)),
+            _ => unreachable!(),
+        }
+    }
+
+    #[test]
+    #[allow(clippy::float_cmp)]
+    fn test_append_rpn_nodes_recursively_1() {
+        use crate::util::codec::number::NumberEncoder;
+        use cop_datatype::FieldTypeTp;
+
+        // Input:
+        // FnD(a, FnA(FnC(b, c, d)), FnA(FnB(e, f))
+        //
+        // Tree:
+        //           FnD
+        // +----------+----------+
+        // a         FnA        FnA
+        //            |          |
+        //           FnC        FnB
+        //        +---+---+      +---+
+        //        b   c   d      e   f
+        //
+        // RPN:
+        // a b c d FnC FnA e f FnB FnA FnD
+
+        let node_fn_a_1 = {
+            // node b
+            let mut node_b = Expr::new();
```

Comment:
Can we extract to a function like this?
```rust
        fn new_expr(exprType: ExprType, fieldTypeTp: FieldTypeTp, childExprs: Vec<Expr>) -> Expr {
            let mut node = Expr::new();
            node.set_tp(exprType);
            node
                .mut_field_type()
                .as_mut_accessor()
                .set_tp(fieldTypeTp);
            for child in childExprs {
                node.mut_children().push(child);
            }
            node
        }
```

Reply:
Yup. Planned to do this (named like KvprotoHelper or TipbHelper) after we merged all executors, providing some commonly used utilities.